tests/test_generate_data.py

```diff
@@ -3,12 +3,464 @@
+        assert "doc_to_choice" in yaml_contents
+        assert "doc_to_target" in yaml_contents
+
+
+def validate_mmlubench_dataset(dataset_file_name):
+    with open(dataset_file_name, encoding="utf-8") as fp:
+        # FIXME: fix the mmlubench pipeline in this test
+        assert fp.readlines() == []
+
+
+def generate_test_samples(yaml_contents):
+    """Convert questions and answers from the taxonomy format into the
+    user/assistant format used by the legacy training methods such as
+    Linux training and MacOS training.
+
+    This mirrors what _gen_test_data() does.
+    """
+    test_samples = []
+    is_knowledge = "document" in yaml_contents
+    for seed_example in yaml_contents["seed_examples"]:
+        if is_knowledge:
+            for qna in seed_example["questions_and_answers"]:
+                test_samples.append(
+                    {
+                        "user": qna["question"]
+                        + "\n"
+                        + seed_example["context"].strip(),
+                        "assistant": qna["answer"].strip(),
+                    }
+                )
+
+        else:
+            # FIXME: handle freeform skills - no context
+            test_samples.append(
+                {
+                    "user": seed_example["question"] + "\n" + seed_example["context"],
+                    "assistant": seed_example["answer"],
+                }
+            )
+    return test_samples
+
+
+def generate_train_samples(yaml_contents):
+    """Generate expected training samples in the user/assistant format
+    used by the legacy training methods such as Linux training and MacOS
+    training.
+
+    Mirroring _noop_llmblock_generate() below, we generate 10 samples
+    per input, and then follow _gen_train_data()'s output format.
+    """
+
+    def add_q(q):
+        return (q + "?") if not "?" in q else q
+
+    train_samples = []
+    is_knowledge = "document" in yaml_contents
+    for seed_example in yaml_contents["seed_examples"]:
+        for i in range(10):
+            if is_knowledge:
+                train_samples.append(
+                    {
+                        "user": seed_example["context"]
+                        + f" (q{i}) "
+                        + add_q(
+                            seed_example["questions_and_answers"][0]["question"].strip()
+                        ),
+                        "assistant": f"(a{i}) "
+                        + seed_example["questions_and_answers"][0]["answer"].strip(),
+                    }
+                )
+            else:
+                # FIXME: handle freeform skills - no context
+                train_samples.append(
+                    {
+                        "user": seed_example["context"]
+                        + f" (q{i}) "
+                        + add_q(seed_example["question"]),
+                        "assistant": f"(a{i}) " + seed_example["answer"],
+                    }
+                )
```

Comment:
at this level of indentation I feel it is cleaner to export this logic to a separate function that can be just called in the loop for each seed example. Maybe something like `process_seed_example(seed_example)` or maybe something more specific if possible.

Also, is there a specific reason why we only want the first 10 Q&As? Apologies if I'm missing something.

Reply:
> at this level of indentation I feel it is cleaner to export this logic to a separate function that can be just called in the loop for each seed example. Maybe something like process_seed_example(seed_example) or maybe something more specific if possible.

I am on the fence about this. Its on the border of being over indented and as the code block is not used anywhere else, I am going to leave it as is for now.

Reply:
> Also, is there a specific reason why we only want the first 10 Q&As? Apologies if I'm missing something.

This is generating 10 examples for training samples, for test purposes. 10 is probably enough for the overall testing of the pipeline.